File: crates/ub-local/src/api/jobs.rs
use crate::config::settings;
use crate::orchestrate::parse_document;
use crate::pipeline::{ensure_output_subdir, export_knowledge_bundle, upload_bundle};
use chrono::Utc;
use serde::Serialize;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use tokio::sync::Semaphore;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize)]
pub struct Job {
    pub id: String,
    pub kind: String,
    pub state: String,
    pub phase: String,
    pub progress: f32,
    pub created_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub error: Option<String>,
    pub result: Option<Value>,
    pub logs: Vec<String>,
}

impl Job {
    pub fn new(kind: &str) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            kind: kind.to_string(),
            state: "pending".into(),
            phase: "queued".into(),
            progress: 0.0,
            created_at: Utc::now().to_rfc3339(),
            started_at: None,
            finished_at: None,
            error: None,
            result: None,
            logs: Vec::new(),
        }
    }

    fn log(&mut self, msg: impl Into<String>) {
        let line = format!("[{}] {}", Utc::now().to_rfc3339(), msg.into());
        self.logs.push(line);
        if self.logs.len() > 2000 {
            let keep = self.logs.split_off(self.logs.len() - 1500);
            self.logs = keep;
        }
    }

    fn set_progress(&mut self, phase: &str, progress: f32, msg: Option<&str>) {
        self.phase = phase.to_string();
        self.progress = progress.clamp(0.0, 100.0);
        if let Some(m) = msg {
            self.log(format!("[{phase} {:.0}%] {m}", self.progress));
        }
    }
}

#[derive(Clone)]
pub struct JobStore {
    inner: Arc<Mutex<HashMap<String, Job>>>,
    limit: Arc<Semaphore>,
}

impl JobStore {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(HashMap::new())),
            limit: Arc::new(Semaphore::new(2)),
        }
    }

    pub fn get(&self, id: &str) -> Option<Job> {
        self.inner.lock().ok()?.get(id).cloned()
    }

    pub fn upsert(&self, job: Job) {
        if let Ok(mut g) = self.inner.lock() {
            g.insert(job.id.clone(), job);
        }
    }

    pub fn mutate<F: FnOnce(&mut Job)>(&self, id: &str, f: F) {
        if let Ok(mut g) = self.inner.lock() {
            if let Some(job) = g.get_mut(id) {
                f(job);
            }
        }
    }

    pub fn start_pipeline(
        &self,
        input: PathBuf,
        track: String,
        mode: String,
        upload: bool,
        skip_mineru: bool,
        org_id: Option<String>,
        document_id: Option<String>,
        filename: Option<String>,
    ) -> Job {
        let mut job = Job::new("pipeline");
        let id = job.id.clone();
        let id_for_return = id.clone();
        self.upsert(job.clone());
        let store = self.clone();
        let permit_sem = self.limit.clone();
        tokio::spawn(async move {
            let _permit = permit_sem.acquire().await.ok();
            store.mutate(&id, |j| {
                j.state = "running".into();
                j.started_at = Some(Utc::now().to_rfc3339());
                j.set_progress("parse", 5.0, Some("starting pipeline"));
            });
            let result = run_pipeline(
                &store,
                &id,
                &input,
                &track,
                &mode,
                upload,
                skip_mineru,
                org_id.as_deref(),
                document_id.as_deref(),
                filename.as_deref(),
            )
            .await;
            store.mutate(&id, |j| {
                j.finished_at = Some(Utc::now().to_rfc3339());
                match result {
                    Ok(v) => {
                        j.state = "done".into();
                        j.result = Some(v);
                        j.set_progress("done", 100.0, Some("finished"));
                    }
                    Err(e) => {
                        j.state = "failed".into();
                        j.phase = "failed".into();
                        j.error = Some(e.to_string());
                        j.log(e.to_string());
                    }
                }
            });
        });
        self.get(&id_for_return).unwrap_or(job)
    }
}

async fn run_pipeline(
    store: &JobStore,
    job_id: &str,
    input: &Path,
    track: &str,
    mode: &str,
    upload: bool,
    skip_mineru: bool,
    org_id: Option<&str>,
    document_id: Option<&str>,
    filename: Option<&str>,
) -> anyhow::Result<Value> {
    let _ = track; // survey track: same export path for now when corpus-like
    let work = ensure_output_subdir(&format!("jobs/{job_id}"));
    store.mutate(job_id, |j| j.set_progress("parse", 10.0, Some("parsing")));
    let corpus = parse_document(input, &work.join("parse"), skip_mineru)?;
    store.mutate(job_id, |j| {
        j.set_progress("embed", 35.0, Some("export + embed"));
    });
    let doc_id = document_id.unwrap_or("local-doc");
    let export_dir = work.join("bundle");
    let store2 = store.clone();
    let jid = job_id.to_string();
    let export = export_knowledge_bundle(
        &corpus,
        &export_dir,
        doc_id,
        filename,
        org_id,
        Some(Box::new(move |done, total| {
            let pct = 35.0 + 50.0 * (done as f32 / total.max(1) as f32);
            store2.mutate(&jid, |j| {
                j.set_progress("embed", pct, Some(&format!("embedded {done}/{total}")));
            });
        })),
    )
    .await?;

    if !upload {
        return Ok(json!({ "export": export, "uploaded": false }));
    }
    store.mutate(job_id, |j| j.set_progress("upload", 90.0, Some("uploading")));
    let uploaded = upload_bundle(&export_dir, mode, org_id, None).await?;
    Ok(json!({ "export": export, "upload": uploaded }))
}

pub async fn health_payload() -> Value {
    let s = settings();
    json!({
        "ok": true,
        "host": s.ub_local_host,
        "port": s.ub_local_port,
        "rag": crate::rag::rag_ready().await,
        "mineru": crate::orchestrate::check_mineru(),
        "deepread": crate::orchestrate::check_deepread(),
        "ssh_configured": !s.ssh_target.trim().is_empty(),
        "org_id": s.survey_org_id,
        "api_url": s.userbank_api_url,
        "work_dir": s.work_dir().display().to_string(),
    })
}

File: src/jobs.py
from __future__ import annotations

import asyncio
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable

from .config import get_settings
from .orchestrate import check_deepread, check_mineru, parse_document
from .pipeline import ensure_output_subdir, export_knowledge_bundle, upload_bundle
from .rag import rag_ready


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass
class Job:
    id: str
    kind: str
    state: str = "pending"
    phase: str = "queued"
    progress: float = 0.0
    created_at: str = field(default_factory=_now)
    started_at: str | None = None
    finished_at: str | None = None
    error: str | None = None
    result: dict[str, Any] | None = None
    logs: list[str] = field(default_factory=list)

    @classmethod
    def new(cls, kind: str) -> Job:
        return cls(id=str(uuid.uuid4()), kind=kind)

    def log(self, msg: str) -> None:
        self.logs.append(f"[{_now()}] {msg}")
        if len(self.logs) > 2000:
            self.logs = self.logs[-1500:]

    def set_progress(self, phase: str, progress: float, msg: str | None = None) -> None:
        self.phase = phase
        self.progress = max(0.0, min(100.0, progress))
        if msg:
            self.log(f"[{phase} {self.progress:.0f}%] {msg}")

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "kind": self.kind,
            "state": self.state,
            "phase": self.phase,
            "progress": self.progress,
            "created_at": self.created_at,
            "started_at": self.started_at,
            "finished_at": self.finished_at,
            "error": self.error,
            "result": self.result,
            "logs": list(self.logs),
        }


class JobStore:
    def __init__(self) -> None:
        self._jobs: dict[str, Job] = {}
        self._lock = asyncio.Lock()
        self._limit = asyncio.Semaphore(2)

    async def get(self, job_id: str) -> Job | None:
        async with self._lock:
            job = self._jobs.get(job_id)
            return job

    def get_sync(self, job_id: str) -> Job | None:
        return self._jobs.get(job_id)

    async def upsert(self, job: Job) -> None:
        async with self._lock:
            self._jobs[job.id] = job

    def upsert_sync(self, job: Job) -> None:
        self._jobs[job.id] = job

    async def mutate(self, job_id: str, fn: Callable[[Job], None]) -> None:
        async with self._lock:
            job = self._jobs.get(job_id)
            if job is not None:
                fn(job)

    def start_pipeline(
        self,
        input_path: Path,
        track: str,
        mode: str,
        upload: bool,
        skip_mineru: bool,
        org_id: str | None,
        document_id: str | None,
        filename: str | None,
    ) -> Job:
        job = Job.new("pipeline")
        self.upsert_sync(job)
        asyncio.create_task(
            self._run_pipeline_task(
                job.id,
                input_path,
                track,
                mode,
                upload,
                skip_mineru,
                org_id,
                document_id,
                filename,
            )
        )
        return job

    def start_upload(
        self,
        bundle_dir: Path,
        mode: str,
        org_id: str | None,
    ) -> Job:
        job = Job.new("upload")
        self.upsert_sync(job)
        asyncio.create_task(self._run_upload_task(job.id, bundle_dir, mode, org_id))
        return job

    async def _run_upload_task(
        self,
        job_id: str,
        bundle_dir: Path,
        mode: str,
        org_id: str | None,
    ) -> None:
        def mark_running(j: Job) -> None:
            j.state = "running"
            j.started_at = _now()

        await self.mutate(job_id, mark_running)
        try:
            result = await upload_bundle(bundle_dir, mode, org_id, None)

            def done(j: Job) -> None:
                j.state = "done"
                j.result = result
                j.phase = "done"
                j.progress = 100.0
                j.finished_at = _now()

            await self.mutate(job_id, done)
        except Exception as exc:  # noqa: BLE001

            def fail(j: Job) -> None:
                j.state = "failed"
                j.error = str(exc)
                j.finished_at = _now()

            await self.mutate(job_id, fail)

    async def _run_pipeline_task(
        self,
        job_id: str,
        input_path: Path,
        track: str,
        mode: str,
        upload: bool,
        skip_mineru: bool,
        org_id: str | None,
        document_id: str | None,
        filename: str | None,
    ) -> None:
        async with self._limit:

            def mark_running(j: Job) -> None:
                j.state = "running"
                j.started_at = _now()
                j.set_progress("parse", 5.0, "starting pipeline")

            await self.mutate(job_id, mark_running)
            try:
                result = await run_pipeline(
                    self,
                    job_id,
                    input_path,
                    track,
                    mode,
                    upload,
                    skip_mineru,
                    org_id,
                    document_id,
                    filename,
                )

                def done(j: Job) -> None:
                    j.state = "done"
                    j.result = result
                    j.set_progress("done", 100.0, "finished")
                    j.finished_at = _now()

                await self.mutate(job_id, done)
            except Exception as exc:  # noqa: BLE001

                def fail(j: Job) -> None:
                    j.state = "failed"
                    j.phase = "failed"
                    j.error = str(exc)
                    j.log(str(exc))
                    j.finished_at = _now()

                await self.mutate(job_id, fail)


async def run_pipeline(
    store: JobStore,
    job_id: str,
    input_path: Path,
    track: str,
    mode: str,
    upload: bool,
    skip_mineru: bool,
    org_id: str | None,
    document_id: str | None,
    filename: str | None,
) -> dict[str, Any]:
    work = ensure_output_subdir(f"jobs/{job_id}")
    await store.mutate(job_id, lambda j: j.set_progress("parse", 10.0, "parsing"))
    if track == "image":
        from .image_ingest import ingest_image_document

        async def on_vision(phase: str, done: int, total: int) -> None:
            pct = 10.0 + 45.0 * (done / max(total, 1))
            await store.mutate(
                job_id,
                lambda j: j.set_progress(
                    "vision", pct, f"{phase} {done}/{total}"
                ),
            )

        corpus = await ingest_image_document(
            input_path, work / "parse", on_vision
        )
        embed_start = 60.0
    else:
        corpus = await asyncio.to_thread(
            parse_document, input_path, work / "parse", skip_mineru
        )
        embed_start = 35.0
    await store.mutate(job_id, lambda j: j.set_progress("embed", embed_start, "export + embed"))
    doc_id = document_id or "local-doc"
    export_dir = work / "bundle"

    async def on_progress(done: int, total: int) -> None:
        pct = embed_start + (90.0 - embed_start) * (done / max(total, 1))
        await store.mutate(
            job_id,
            lambda j: j.set_progress("embed", pct, f"embedded {done}/{total}"),
        )

    export = await export_knowledge_bundle(
        corpus,
        export_dir,
        doc_id,
        filename,
        org_id,
        on_progress,
    )

    if not upload:
        return {"export": export, "uploaded": False}
    await store.mutate(job_id, lambda j: j.set_progress("upload", 90.0, "uploading"))
    uploaded = await upload_bundle(export_dir, mode, org_id, None)
    return {"export": export, "upload": uploaded}


async def health_payload() -> dict[str, Any]:
    s = get_settings()
    return {
        "ok": True,
        "host": s.ub_local_host,
        "port": s.ub_local_port,
        "rag": await rag_ready(),
        "mineru": check_mineru(),
        "deepread": check_deepread(),
        "ssh_configured": bool(s.ssh_target.strip()),
        "org_id": s.survey_org_id,
        "api_url": s.userbank_api_url,
        "work_dir": str(s.work_dir()),
    }
